### ISSUE TITLE
Structural plasticity delivers spikes via connections that were not present at the time of the spike

### ISSUE THREAD

**Issue by jakobj** [nest/nest-simulator#571]
The `update_` function in NEST contains the following while loop (severely simplified, see https://github.com/nest/nest-simulator/blob/master/nestkernel/simulation_manager.cpp#L588 ff):
```c++
    do
    {
      if ( kernel().sp_manager.is_structural_plasticity_enabled()
        && ( clock_.get_steps() + from_step_ )
            % kernel().sp_manager.get_structural_plasticity_update_interval()
          == 0 )
      {
        [...] // update of structural plasticity
      }

      if ( from_step_ == 0 ) // deliver only at beginning of slice
      {
        kernel().event_delivery_manager.deliver_events( thrd );
        [...]
      }
      [...]

#pragma omp master
      {
        // gather only at end of slice
        if ( to_step_ == kernel().connection_manager.get_min_delay() )
          kernel().event_delivery_manager.gather_events( true );
        [...]
      }
      [...]
    } while ( ( to_do_ != 0 ) && ( !terminate_ ) );
```

As one can deduce from the comments, spikes are always "gathered", i.e., communicated among all processes at the end of a slice, while they are delivered at the beginning of the next slide. However, due to the order of operations in the while loop, structural plasticity can create connections that were not present when spikes were gathered in the previous step. This can lead to the following situation in two subsequent slices:
1. neurons 1 and 2 are unconnected
2. neuron 1 spikes
3. structural plasticity creates a connection from neuron 1 to neuron 2
4. a spike is delivered to neuron 2

At least for me, this is quite counterintuitive, since (for some reason) I expect spikes only to be delivered via connections *that were present at the time of the spike*. However, my intuition might be wrong, so I would like to discuss this issue.
In case we agree this is undesired behaviour, an easy fix (that I have already prepare locally) is to move `deliver` above the structural plasticity update.
@sdiazpier @suku248 @heplesser, please have a look at this.

**Comment by heplesser** [comment]
@jakobj The mysterious ways of network simulation ... I agree with your analysis in principle. One could argue, though, that structural plasticity occurs at a much large time scale than spiking, and that time itself thus should be considered coarse-grained with respect to synapse creation and deletion. Let us assume SP is updated very 100 ms. In a real network, new synapses are created at any time, not at multiples of 100 ms. Thus, a synapse created in the simulation at 200 ms might represent a synapse created in reality at 185 ms, or at 205 ms. Based on this, I would argue that it does not matter for the biological relevance of SP simulations whether spikes can pass through "future" synapses or not: in biology, the synapse might already have been there.

But from the point of minimal surprise for the NEST user, it certainly would make sense to prohibit spikes sent through synapses yet to be created. But I fear your fix is too simple in two ways:

- Currently, with delivery after SP, no spike is transmitted through synapses that will be deleted. With your fix, that would change. A complete fix would require deleting synapses before delivery and creating them after.
- Which time point is relevant? Your change puts it to the rather arbitrary time at which spike exchange occurs. Furthermore, it ignores delay, which may be long. I would argue that the axonal delay is relevant, since a synapse deleted after a spike has already entered the cell would change things. Again, this point I think rather shows that these are phenomena on different time scales.

**Comment by jakobj** [comment]
Thanks for you commen @heplesser. I agree that it is difficult to argue about biological plausibility with this coarse grained update of synaptic elements, so the technical/user experience view should take precedence.
In this regard, I disagree with your first point: from the point of the user, spikes should be delivered via all and only those synapses that are present when the presynaptic neuron was active. Why should we treat synapses that will be deleted in the next sp step differently, and disregard their spikes is a single iteration of the whole update loop, which happens to coincide with the time scale of sp?
I am uncertain about your second point though. However, I do not feel that I am choosing some arbitrary time point. Spikes are generated and delivered as usual, with structural plasiicity acting similar to a user pausing and restarting the simulation.

**Comment by heplesser** [comment]
@jakobj I would suggest that we discuss this during one of the next developer meetings.

**Comment by heplesser** [comment]
To be discussed in Developer VC 29 Oct 2018.

**Comment by heplesser** [comment]
@sdiazpier Could you share your thoughts on this issue?

**Comment by stinebuu** [comment]
Fixed by #1395.
